File: blog/src/main/java/com/github/blog/controller/advice/BaseExceptionHandler.java
package com.github.blog.controller.advice;

import com.github.blog.service.exception.ExceptionEnum;
import com.github.blog.service.exception.impl.CustomException;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.AuthenticationException;
import org.springframework.validation.FieldError;
import org.springframework.validation.ObjectError;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import java.util.ArrayList;
import java.util.List;

/**
 * @author Raman Haurylau
 */
@RestControllerAdvice
public class BaseExceptionHandler {

    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<ValidationErrorResponse> handleValidationExceptions(MethodArgumentNotValidException e) {
        List<Violation> violations = new ArrayList<>();

        for (ObjectError error : e.getBindingResult().getAllErrors()) {
            if (error instanceof FieldError) {
                violations.add(new Violation(((FieldError) error).getField(), error.getDefaultMessage()));
            } else {
                violations.add(new Violation(error.getObjectName(), error.getDefaultMessage()));
            }
        }

        return ResponseEntity
                .status(e.getStatusCode())
                .body(new ValidationErrorResponse("VALIDATION_FAILED", violations, System.currentTimeMillis()));
    }

    @ExceptionHandler(AuthenticationException.class)
    public ResponseEntity<ErrorResponse> handleCustomEntityException() {
        return ResponseEntity
                .status(HttpStatus.UNAUTHORIZED)
                .body(new ErrorResponse(HttpStatus.UNAUTHORIZED.name(), ExceptionEnum.AUTHENTICATION_FAILED.getMessage(), System.currentTimeMillis()));
    }

    @ExceptionHandler(CustomException.class)
    public ResponseEntity<ErrorResponse> handleCustomEntityException(CustomException e) {
        return ResponseEntity
                .status(e.getExceptionEnum().getStatus())
                .body(new ErrorResponse(e.getExceptionEnum().name(), e.getExceptionEnum().getMessage(), System.currentTimeMillis()));
    }

    @ExceptionHandler(RuntimeException.class)
    public ResponseEntity<ErrorResponse> handleException(RuntimeException e) {
        return ResponseEntity
                .status(HttpStatus.BAD_REQUEST)
                .body(new ErrorResponse(HttpStatus.BAD_REQUEST.name(), e.getMessage(), System.currentTimeMillis()));
    }

    @ExceptionHandler(Exception.class)
    public ResponseEntity<ErrorResponse> handleException(Exception e) {
        return ResponseEntity
                .status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(new ErrorResponse(HttpStatus.INTERNAL_SERVER_ERROR.name(), e.getMessage(), System.currentTimeMillis()));
    }
}

File: api-gateway/src/main/kotlin/com/github/api_gateway/controller/advice/BaseExceptionHandler.kt
package com.github.api_gateway.controller.advice

import com.github.api_gateway.service.exception.ExceptionEnum
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.core.AuthenticationException
import org.springframework.validation.FieldError
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice

/**
 * @author Raman Haurylau
 */
@RestControllerAdvice
class BaseExceptionHandler {

    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleValidationException(e: MethodArgumentNotValidException): ResponseEntity<ValidationErrorResponse> {
        val violations: MutableList<Violation> = mutableListOf()

        e.bindingResult.allErrors.forEach {
            if (it is FieldError) {
                violations.add(Violation(it.field, it.defaultMessage))
            } else {
                violations.add(Violation(it.objectName, it.defaultMessage))
            }
        }

        return ResponseEntity
            .status(e.statusCode)
            .body(ValidationErrorResponse("VALIDATION_FAILED", violations, System.currentTimeMillis()))
    }

    @ExceptionHandler(AuthenticationException::class)
    fun handleCustomEntityException(): ResponseEntity<ErrorResponse> {
        return ResponseEntity
            .status(HttpStatus.UNAUTHORIZED)
            .body(ErrorResponse(HttpStatus.UNAUTHORIZED.name, ExceptionEnum.AUTHENTICATION_FAILED.message, System.currentTimeMillis()))
    }

    @ExceptionHandler(RuntimeException::class)
    fun handleException(e: RuntimeException): ResponseEntity<ErrorResponse> {
        return ResponseEntity
            .status(HttpStatus.BAD_REQUEST)
            .body(ErrorResponse(HttpStatus.BAD_REQUEST.name, e.message, System.currentTimeMillis()))
    }

    @ExceptionHandler(Exception::class)
    fun handleException(e: Exception): ResponseEntity<ErrorResponse> {
        return ResponseEntity
            .status(HttpStatus.INTERNAL_SERVER_ERROR)
            .body(ErrorResponse(HttpStatus.INTERNAL_SERVER_ERROR.name, e.message, System.currentTimeMillis()))
    }
}